File: tests/solutions_cpp/minimum_window_substring.cpp
#include "json_helper.h"

class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) return "";
        unordered_map<char, int> need;
        for (char c : t) need[c]++;
        int missing = t.size();
        int bestLeft = 0, bestLen = INT_MAX;
        int left = 0;
        for (int right = 0; right < (int)s.size(); right++) {
            char c = s[right];
            if (need[c] > 0) missing--;
            need[c]--;
            if (missing == 0) {
                while (need[s[left]] < 0) {
                    need[s[left]]++;
                    left++;
                }
                if (right - left + 1 < bestLen) {
                    bestLen = right - left + 1;
                    bestLeft = left;
                }
                need[s[left]]++;
                missing++;
                left++;
            }
        }
        return bestLen == INT_MAX ? "" : s.substr(bestLeft, bestLen);
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    string s = to_string_val(args[0]);
    string t = to_string_val(args[1]);
    string result = sol.minWindow(s, t);
    cout << result << endl;
    return 0;
}

File: tests/solutions_js/minimum_window_substring.js
class Solution {
    minWindow(s, t) {
        if (!t || !s) return "";
        const need = {};
        for (const c of t) need[c] = (need[c] || 0) + 1;
        let missing = t.length;
        let bestLeft = 0, bestRight = Infinity;
        let left = 0;
        for (let right = 0; right < s.length; right++) {
            const char = s[right];
            if ((need[char] || 0) > 0) missing--;
            need[char] = (need[char] || 0) - 1;
            if (missing === 0) {
                while (need[s[left]] < 0) {
                    need[s[left]]++;
                    left++;
                }
                if (right - left < bestRight - bestLeft) {
                    bestLeft = left;
                    bestRight = right;
                }
                need[s[left]]++;
                missing++;
                left++;
            }
        }
        return bestRight !== Infinity ? s.slice(bestLeft, bestRight + 1) : "";
    }
}
